src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,256 @@
+                return;
+
+            var model = await _updateDataModelQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
+            if (model is null)
+                return;
+
+            // Switch to the UI thread to get the current caret point and latest active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return;
+
+            var caretPoint = activeTextView.GetCaretPoint(activeTextView.TextBuffer);
+            if (!caretPoint.HasValue)
+                return;
+
+            var currentSnapshot = activeTextView.TextSnapshot;
+            var caretPosition = caretPoint.Value.Position;
+
+            // Switch to the threadpool to determine which node is currently selected and which node to select (if they exist).
+            await TaskScheduler.Default;
+
+            var currentlySelectedSymbol = DocumentOutlineHelper.GetCurrentlySelectedNode(model.DocumentSymbolItems);
+            var symbolToSelect = DocumentOutlineHelper.GetDocumentNodeToSelect(model.DocumentSymbolItems, model.OriginalSnapshot, currentSnapshot, caretPosition);
+
+            // Switch to the UI thread to update the view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            // Unselects the currently selected symbol. This is required in case the current caret position is not in range of any symbols
+            // (symbolToSelect will be null) so that no symbols in the tree are highlighted.
+            if (currentlySelectedSymbol is not null)
+                currentlySelectedSymbol.IsSelected = false;
+
+            if (symbolToSelect is not null)
+                symbolToSelect.IsSelected = true;
+
+            var expansion = expansionOption.First();
+            if (expansion is not ExpansionOption.NoChange)
+                DocumentOutlineHelper.SetIsExpanded(model.DocumentSymbolItems, expansion);
+
+            symbolTree.ItemsSource = model.DocumentSymbolItems;
+        }
+
+        /// <summary>
+        /// Starts a new task to select code when a symbol node is clicked.
+        /// </summary>
+        private void StartJumpToContentTask(DocumentSymbolItem symbol)
+        {
+            _jumpToContentQueue.AddWork(symbol);
+        }
+
+        /// <summary>
+        /// Given a DocumentSymbolItem, moves the caret to its position in the latest active text view.
+        /// </summary>
+        private async ValueTask JumpToContentAsync(ImmutableSegmentedList<DocumentSymbolItem> symbol, CancellationToken cancellationToken)
+        {
+            if (symbol.IsDefault || symbol.IsEmpty)
+                return;
+
+            var model = await _computeDataModelQueue.WaitUntilCurrentBatchCompletesAsync().ConfigureAwait(false);
```

Comment:
Why the wait here? If I've double clicked an item, don't we already have the info at that point to do the navigation? Or could this be a different model and things could get out of sync?